flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java

```diff
@@ -0,0 +1,180 @@
+     *     the factory class was not found or the factory could not be instantiated
+     * @throws IllegalConfigurationException May be thrown by the CheckpointStorageFactory when
+     *     creating / configuring the checkpoint storage in the factory
+     * @throws IOException May be thrown by the CheckpointStorageFactory when instantiating the
+     *     checkpoint storage
+     */
+    public static CheckpointStorage loadCheckpointStorageFromConfig(
+            ReadableConfig config, ClassLoader classLoader, @Nullable Logger logger)
+            throws IllegalStateException, DynamicCodeLoadingException, IOException {
+
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(classLoader, "classLoader");
+
+        final String storageName = config.get(CheckpointingOptions.CHECKPOINT_STORAGE);
+        if (storageName == null) {
+            return null;
+        }
+
+        switch (storageName.toLowerCase()) {
+            case JOB_MANAGER_STORAGE_NAME:
+                throw new IllegalStateException(
+                        "JobManagerCheckpointStorage is not yet implemented");
+
+            case FILE_SYSTEM_STORAGE_NAME:
+                throw new IllegalStateException(
+                        "FileSystemCheckpointStorage is not yet implemented");
+
+            default:
+                if (logger != null) {
+                    logger.info("Loading state backend via factory {}", storageName);
+                }
+
+                CheckpointStorageFactory<?> factory;
+                try {
+                    @SuppressWarnings("rawtypes")
+                    Class<? extends CheckpointStorageFactory> clazz =
+                            Class.forName(storageName, false, classLoader)
+                                    .asSubclass(CheckpointStorageFactory.class);
+
+                    factory = clazz.newInstance();
+                } catch (ClassNotFoundException e) {
+                    throw new DynamicCodeLoadingException(
+                            "Cannot find configured state backend factory class: " + storageName,
+                            e);
+                } catch (ClassCastException | InstantiationException | IllegalAccessException e) {
+                    throw new DynamicCodeLoadingException(
+                            "The class configured under '"
+                                    + CheckpointingOptions.CHECKPOINT_STORAGE.key()
+                                    + "' is not a valid checkpoint storage factory ("
+                                    + storageName
+                                    + ')',
+                            e);
+                }
+
+                return factory.createFromConfig(config, classLoader);
+        }
+    }
+
+    public static CheckpointStorage fromApplicationOrConfigOrDefault(
+            @Nullable CheckpointStorage fromApplication,
```

Comment:
Flink's style guide says not to use `Optional` for function arguments. 
https://flink.apache.org/contributing/code-style-and-quality-java.html#java-optional

Reply:
Do you mean
> Do not use Optional as a function argument, instead either overload the method or use the Builder pattern for the set of function arguments.

I think the intention is to avoid arguments that can be empty/null. According to it the method should be refactored.

But I think the very purpose of this method is to chose which arguments to use. And for that I'd use `Optional`.